### ISSUE TITLE
Sending long lines terminated by "\r" to terminal hangs Vim

### ISSUE THREAD

**Issue by lifepillar** [vim/vim#1985]
To reproduce, source this file:

```vim
let lines = ['xxxxxx',
      \ 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
      \ 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
      \ 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
      \ 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
      \ 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
      \ 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
      \ 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
      \ 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
      \ 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
      \ 'xxxxxxxxxxxxxxxxxx']

let bn = term_start(&shell)
call term_sendkeys(bn, join(lines, "\<cr>")."\<cr>")
```

Using Vim 8.0.946 on macOS. I am able to reproduce the problem in Terminal.app and iTerm2.

Also note that if I close the terminal app's window (not Vim's) to force quit Vim, the shell process started by Vim terminal is not killed.

**Comment by mattn** [comment]
I don't reproduce this on Linux and Windows both.

**Comment by lifepillar** [comment]
On macOS, I can consistently reproduce it even with `vim -u NONE -N`. Vim is installed with Homebrew, btw.

**Comment by lifepillar** [comment]
If I try with a shorter text, it works. It might be a buffer overflow problem. Please try with a longer text.

**Comment by nuko8** [comment]
It follows from the script that all those `x`'s are passed to the command line of the shell all at once; since there's no separator, they and the `\r`'s between them are all treated as a single command (i.e., the 0th argument), and hence, are passed to the first argument of an `exec*()` function as a single, long string.

The first argument of `exec*()` functions is a path, and there's a limitation on its length imposed by the OS: The number is 1024 (`PATH_MAX`).

Counting the number of `x`'s in the example, I got 1,290.   So it looks to me that the OP is asking too much of his OS...

**Comment by lifepillar** [comment]
I am not familiar with the internal implementation of `term_sendkeys()`, but why should it call `exec*()` at all? As far as I understand, it just forwards the keys to the terminal (so, I assume that sending `\<cr>` is like pressing Enter at the shell prompt—please correct if I am wrong). Anyway, how is one supposed to send multiple commands or a multi-line command to the shell? Should I call `term_sendkeys()` once for each line?

To be clear, my use case is the following: I want to visually select a complex SQL query in a Vim buffer and send the selection to the terminal to be executed by a SQL client.

In any case, clearly it would be better if Vim failed more gracefully.

**Comment by lifepillar** [comment]
>since there's no separator, they and the \r's between them are all treated as a single command (i.e., the 0th argument), and hence, are passed to the first argument of an exec*() function as a single, long string.

Ok, I have misunderstood: here you mean the shell's `exec*()` function. But anyway, what you describe is not the behaviour I see. If I send `ls\<cr>pwd`, the shell first executes `ls`, and then executes `pwd`. I don't see how this is interpreted as a single command.

>Counting the number of x's in the example, I got 1,290. So it looks to me that the OP is asking too much of his OS...

When I send 2000 `x`'s to the shell, I just get `command not found`.

**Comment by nuko8** [comment]
> If I send ls\<cr>pwd, the shell first executes ls, and then executes pwd. I don't see how this is interpreted as a single command.

If the length of the given string is short enough for the system to handle, then interpreting `\r`'s between commands makes sense.  On the other hand, if the length were too long, then it wouldn't make sense to check the existence of`\r`'s to divide the string into tokens or commands (In other words, after the check, what is the purpose of the system to make use of `\r`'s?)  Probably, that's the reason why `ls\<cr>pwd` works. 

> When I send 2000 x's to the shell, I just get command not found.

In addition to that, `zsh` gives `zsh: command too long: xxxx...` while `sh` simply echoes the given command.

**Comment by lifepillar** [comment]
This snippet:

```vim
let bn = term_start(&shell)
for c in ['a','b','c','d','e','f','g','h','i','j','k']
  call term_sendkeys(bn, repeat(i, 10)."\<cr>")
endfor
```

produces the following output:

```
kkkkkkkkkk
kkkkkkkkkk
kkkkkkkkkk
kkkkkkkkkk
kkkkkkkkkk
kkkkkkkkkk
kkkkkkkkkk
kkkkkkkkkk
kkkkkkkkkk
kkkkkkkkkk
kkkkkkkkkk
me at MyHost in ~
]kkkkkkkkkk
bash: kkkkkkkkkk: command not found
✗ me at MyHost in ~
]kkkkkkkkkk
bash: kkkkkkkkkk: command not found
✗ me at MyHost in ~
]kkkkkkkkkk
bash: kkkkkkkkkk: command not found
✗ me at MyHost in ~
]kkkkkkkkkk
bash: kkkkkkkkkk: command not found
✗ me at MyHost in ~
]kkkkkkkkkk
bash: kkkkkkkkkk: command not found
✗ me at MyHost in ~
]kkkkkkkkkk
bash: kkkkkkkkkk: command not found
✗ me at MyHost in ~
]kkkkkkkkkk
bash: kkkkkkkkkk: command not found
✗ me at MyHost in ~
]kkkkkkkkkk
bash: kkkkkkkkkk: command not found
✗ me at MyHost in ~
]kkkkkkkkkk
bash: kkkkkkkkkk: command not found
✗ me at MyHost in ~
]kkkkkkkkkk
bash: kkkkkkkkkk: command not found
✗ me at MyHost in ~
]kkkkkkkkkk
bash: kkkkkkkkkk: command not found
✗ me at MyHost in ~
```

What's going on here?

**Comment by nuko8** [comment]
> What's going on here?

Hmm, first, the script fails to give the definition of the variable `i`, the first argument of `repeat()`. That explains why you got many `k`'s.

Last but not least, an excessively long argument to `term_sendkeys()` makes Vim hang due to a limitation imposed by the OS, but if the argument is divided into smaller tokens of appropriate granularity in such a way that each of those tokens is executable as a shell command, and if each of them are passed sequentially to `term_sendkeys()` using a loop, then Vim won't hang.

**Comment by lifepillar** [comment]
>Hmm, first, the script fails to give the definition of the variable i, the first argument of repeat(). That explains why you got many k's.

Yes, that's a typo. But I would not expect the first eleven lines, which appear even with the correct script.

**Comment by lifepillar** [comment]
I can still make Vim hang if I send the lines one by one:

```vim
let lines = ['xxxxxx',
      \ 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
      \ 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
      \ 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
      \ 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
      \ 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
      \ 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
      \ 'xxxxxxxxxxxxxxxxxx']

let bn = term_start(&shell)
for line in lines
  call term_sendkeys(bn, line."\<cr>")
endfor
```

If I don't append `"\<cr>"`, both this version and the one I started this issue with work (although I still see the problem with spurious lines at the beginning, mentioned in my previous comment).

**Comment by lifepillar** [comment]
>I can still make Vim hang if I send the lines one by one:

Calling `term_wait()` after sending each line fixes the problem for me.

**Comment by nuko8** [comment]
> Calling term_wait() after sending each line fixes the problem for me.

Personally, that reminds me of a test failure of Test_terminal_env() , the issue of which was fixed by Patch 8.0.0939 (51c23687a).  Also, there's [a thread](https://groups.google.com/forum/#!topicsearchin/vim_dev/Test_terminal_env;context-place=forum/vim_dev/vim_dev/QQjdJ0KfQIE) relevant to that topic.  In particular, please pay attention to Bram's comment on `term_wait()` in the thread.  That could make your solution more robust.

[EDIT: Follow-up]
By "the thread", I meant 
https://groups.google.com/forum/#!topic/vim_dev/QQjdJ0KfQIE

**Comment by lifepillar** [comment]
`WaitFor()` and `term_wait()` appear functionally equivalent in terminal Vim, in my limited testing. In MacVim, though, `term_wait()` effectively causes the shell to exit, while with `WaitFor()` MacVim appears to hang as soon as I send a few relatively long lines each one terminated by `"\<cr>"`.

MacVim is based on Vim 8.0.891.

**Comment by brammool** [comment]
Lifepillar wrote:

> This snippet:
> 
> ```vim
> let bn = term_start(&shell)
> for c in ['a','b','c','d','e','f','g','h','i','j','k']
>   call term_sendkeys(bn, repeat(i, 10)."\<cr>")
> endfor
> ```
> 
> produces the following output:
> 
> ```
> kkkkkkkkkk
> kkkkkkkkkk
> kkkkkkkkkk
> kkkkkkkkkk
> kkkkkkkkkk
> kkkkkkkkkk
> kkkkkkkkkk
> kkkkkkkkkk
> kkkkkkkkkk
> kkkkkkkkkk
> kkkkkkkkkk
> me at MyHost in ~
> ]kkkkkkkkkk
> bash: kkkkkkkkkk: command not found
> ✗ me at MyHost in ~
> ]kkkkkkkkkk
> bash: kkkkkkkkkk: command not found
> ✗ me at MyHost in ~
> ]kkkkkkkkkk
> bash: kkkkkkkkkk: command not found
> ✗ me at MyHost in ~
> ]kkkkkkkkkk
> bash: kkkkkkkkkk: command not found
> ✗ me at MyHost in ~
> ]kkkkkkkkkk
> bash: kkkkkkkkkk: command not found
> ✗ me at MyHost in ~
> ]kkkkkkkkkk
> bash: kkkkkkkkkk: command not found
> ✗ me at MyHost in ~
> ]kkkkkkkkkk
> bash: kkkkkkkkkk: command not found
> ✗ me at MyHost in ~
> ]kkkkkkkkkk
> bash: kkkkkkkkkk: command not found
> ✗ me at MyHost in ~
> ]kkkkkkkkkk
> bash: kkkkkkkkkk: command not found
> ✗ me at MyHost in ~
> ]kkkkkkkkkk
> bash: kkkkkkkkkk: command not found
> ✗ me at MyHost in ~
> ]kkkkkkkkkk
> bash: kkkkkkkkkk: command not found
> ✗ me at MyHost in ~
> ```
> 
> What's going on here?

The keys are sent to the terminal before the terminal has initialized.
The tty echoes the keys, they will be buffered.  Once the shell is ready
it outputs the prompt and starts reading the text.  In your case it
appears the shell doesn't echo the command, when I try it it does get
echo'ed.

Perhaps when you send long text all at once before the shell is reading
them, the write to the pty blocks.  Then when the terminal starts
outputting text to the terminal, Vim is blocked and doesn't read it.  If
that buffer also gets filled up then we have a deadlock.

Can you use ch_logfile() to get a lof file, so you can check that this
is what happens?  You may need to add some ch_log() statements in the
code to narrow it down, e.g. put these in channel_send():

ch_log(channel, "Writing %d bytes", len);
  if (part == PART_SOCK)
	res = sock_write(fd, (char *)buf, len);
    else
	res = fd_write(fd, (char *)buf, len);
ch_log(channel, "Done writing: result: %d", res);

For me it looks like it sends up to 4096 characters before the shell
prompt is displayed.  Hmm, let's try more than that in one line:


	let bn = term_start('bash')
	for c in ['a','b','c','d','e','f','g','h','i','j','k']
	  call term_sendkeys(bn, repeat(c, 5000)."\<cr>")
	endfor

Yep, it hangs and blocks in the write() call.  I'll see if we can make
that non-blocking.


-- 
ARTHUR:    Be quiet!  I order you to shut up.
OLD WOMAN: Order, eh -- who does he think he is?
ARTHUR:    I am your king!
OLD WOMAN: Well, I didn't vote for you.
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lifepillar** [comment]
Do I need anything special to enable the terminal in a Vim build? I have edited `feature.h` and tried `./configure --with-features=huge`, but the executable does not have any `term_*()` function.

**Comment by nuko8** [comment]
> Do I need anything special to enable the terminal in a Vim build? 

Yes, you need to add `--enable-terminal`.  None of the feature sets does not include the option by default yet.

**Comment by lifepillar** [comment]
@brammool This is what I have logged using your example:

```vim
==== start log session ====
  0.000720 : Starting job: bash
  0.000831 on 0: Created channel
  0.001696 on 0: writing out to buffer '!bash'
  0.001771 on 0: writing err to buffer '!bash'
  0.001853 : ioctl(TIOCSWINSZ) success
  0.002266 SEND on 0: 'a'
  0.002287 on 0: Writing 1 bytes
  0.002303 on 0: Done writing: result: 1
  0.002311 SEND on 0: 'a'
  0.002318 on 0: Writing 1 bytes
  0.002328 on 0: Done writing: result: 1
  […]
  0.281660 SEND on 0: 'i'
  0.281662 on 0: Writing 1 bytes
  0.281664 on 0: Done writing: result: 1
  0.281666 SEND on 0: 'i'
  0.281668 on 0: Writing 1 bytes
 44.173436 on 0: Done writing: result: -1
 44.173461 SEND on 0: 'i'
 44.173467 on 0: Writing 1 bytes
```

At ~44s it's when I interrupted Vim.

My original script produces a similar log.

**Comment by brammool** [comment]
Fixed by 8.0.0957
